// @fal-ai/client/src/middleware.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TARGET_URL_HEADER = void 0;
exports.withMiddleware = withMiddleware;
exports.withProxy = withProxy;
/**
 * Setup a execution chain of middleware functions.
 *
 * @param middlewares one or more middleware functions.
 * @returns a middleware function that executes the given middlewares in order.
 */
function withMiddleware(...middlewares) {
    const isDefined = (middleware) => typeof middleware === "function";
    return (config) => __awaiter(this, void 0, void 0, function* () {
        let currentConfig = Object.assign({}, config);
        for (const middleware of middlewares.filter(isDefined)) {
            currentConfig = yield middleware(currentConfig);
        }
        return currentConfig;
    });
}
exports.TARGET_URL_HEADER = "x-fal-target-url";
function shouldProxy(when) {
    const env = {
        isBrowser: typeof window !== "undefined" && typeof window.document !== "undefined",
    };
    if (typeof when === "function") {
        return when(env);
    }
    if (when === "always") {
        return true;
    }
    return env.isBrowser;
}
function withProxy(config) {
    return (requestConfig) => {
        // if x-fal-target-url is already set, we skip it to keep the middleware
        // re-entrant (e.g. when composed with other proxy-like middlewares)
        if (requestConfig.headers && exports.TARGET_URL_HEADER in requestConfig.headers) {
            return Promise.resolve(requestConfig);
        }
        if (!shouldProxy(config.when)) {
            return Promise.resolve(requestConfig);
        }
        return Promise.resolve(Object.assign(Object.assign({}, requestConfig), { url: config.targetUrl, headers: Object.assign(Object.assign({}, (requestConfig.headers || {})), { [exports.TARGET_URL_HEADER]: requestConfig.url }) }));
    };
}
